lte/gateway/c/session_manager/SessionStateEnforcer.cpp

```diff
@@ -857,4 +860,225 @@ void SessionStateEnforcer::set_pdr_attributes(
+                   << " rx bytes for rule " << record.rule_id();
+    }
+
+    session->add_rule_usage(
+        record.rule_id(), record.rule_version(), record.bytes_tx(),
+        record.bytes_rx(), record.dropped_tx(), record.dropped_rx(),
+        &session_update[imsi][session_id]);
+  }
+  if (records.records().size() > 0) {
+    MLOG(MINFO) << "Received stats for " << sessions_with_reporting_flows.size()
+                << " active sessions and " << dead_sessions_to_cleanup.size()
+                << " stale sessions";
+  }
+  complete_termination_for_released_sessions(
+      session_map, sessions_with_reporting_flows, session_update);
+  cleanup_dead_sessions(dead_sessions_to_cleanup);
+  return;
+}
+
+void SessionStateEnforcer::complete_termination_for_released_sessions(
+    SessionMap& session_map,
+    std::unordered_set<ImsiAndSessionID> sessions_with_reporting_flows,
+    SessionUpdate& session_update) {
+  // Iterate through sessions and notify that report has finished. Terminate any
+  // sessions that can be terminated.
+  std::vector<ImsiAndSessionID> sessions_to_terminate;
+  for (const auto& session_pair : session_map) {
+    const std::string imsi = session_pair.first;
+    for (const auto& session : session_pair.second) {
+      const std::string session_id = session->get_session_id();
+      // If we did not receive a rule record for the session, this means
+      // PipelineD has reported all usage for the session
+      auto imsi_and_session_id = ImsiAndSessionID(imsi, session_id);
+      if (session->get_state() == RELEASE &&
+          sessions_with_reporting_flows.find(imsi_and_session_id) ==
+              sessions_with_reporting_flows.end()) {
+        sessions_to_terminate.push_back(imsi_and_session_id);
+      }
+    }
+  }
+  // Do the actual termination in a separate loop since this can modify the
+  // session map structure
+  for (const auto& pair : sessions_to_terminate) {
+    const auto &imsi = pair.first, &session_id = pair.second;
+    m5g_complete_termination(session_map, imsi, session_id, session_update);
+  }
+
+  return;
+}
+
+void SessionStateEnforcer::cleanup_dead_sessions(
+    const RuleRecordSet dead_sessions_to_cleanup) {
+  if (dead_sessions_to_cleanup.size() == 0) {
+    return;
+  }
+  if (!CLEANUP_DANGLING_FLOWS) {
+    MLOG(MWARNING) << "Not cleaning up " << dead_sessions_to_cleanup.size()
+                   << " dangling sessions in PipelineD due to "
+                      "'cleanup_all_dangling_flows: false' in sessiond.yml";
+    return;
+  }
+  MLOG(MINFO) << "Deactivating flows for " << dead_sessions_to_cleanup.size()
+              << " dangling sessions in PipelineD";
+  for (const RuleRecord& record : dead_sessions_to_cleanup) {
+    deactivate_flows_for_termination(
+        record.sid(), record.ue_ipv4(), record.ue_ipv6(), record.teid());
+  }
+
+  return;
+}
+
+void SessionStateEnforcer::deactivate_flows_for_termination(

```

Comment:
This is a very similar from [pipelined client](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Auri200%2Fmagma.git&branch=feg_merge_credit_reports_into_single_crru&file=lte%2Fgateway%2Fc%2Fsession_manager%2FPipelinedClient.cpp&editor=JetBrains&version=v1.2.1&utm_product_name=CLion&utm_product_version=2021.1.2&start_row=328&start_col=37&end_row=328&end_col=37) 

This code, if different than existing, should go on pipelined client

Reply:
The functionality is different.
The "SessionStateEnforcer::deactivate_flows_for_termination" is filling up the SessionState::SessionInfo structure and calling to "pipelined_client_->set_upf_session" as similar to create/delete/INACTIVE.

Reply:
it looks like the link I included was bad. I fixed it and include more hre. Please see how pipelined client works and see how we host there the code to build the request too.

See how we have different functions creating different request. For 5g I only see `set_upf_session` but for 4g i see different functions that build specific request based on different arguments


For example [deactivate_flows_for_rules](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Auri200%2Fmagma.git&branch=chore_federated_lte_test_1&file=lte%2Fgateway%2Fc%2Fsession_manager%2FPipelinedClient.cpp&editor=JetBrains&version=v1.2.1&utm_product_name=CLion&utm_product_version=2021.1.2&start_row=344&start_col=45&end_row=344&end_col=45) and [deactivate_flows_for_rules_for_termination](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Auri200%2Fmagma.git&branch=chore_federated_lte_test_1&file=lte%2Fgateway%2Fc%2Fsession_manager%2FPipelinedClient.cpp&editor=JetBrains&version=v1.2.1&utm_product_name=CLion&utm_product_version=2021.1.2&start_row=328&start_col=35&end_row=328&end_col=35) they do the same:
1. build a request based on some inputs
2. run `deactivate_flows(req)`

Reply:
For 4g, activate/deactivate flow requests are going by grpc (sessiond-pipelined) and gtp tunnel creation/deletion requests are going by OAI interface (MME-Pipelined).

For 5g, gtp tunnel + activate/deactivate flow requests are going by grpc (sessiond-pipelined).

As per the design proposed by Ulas, the only one grpc method will be required to implement for sending all active sessions information to UPF (Just like snapshot of all active sessions).

Please find the attached ppt for reference.
[Discussion Material for Meeting 06102020 UPF-SMF GRPC Interface (1).pptx](https://github.com/magma/magma/files/7293596/Discussion.Material.for.Meeting.06102020.UPF-SMF.GRPC.Interface.1.pptx)